src/java/org/apache/nutch/crawl/CrawlDb.java

```diff
@@ -129,17 +129,23 @@ public void update(Path crawlDb, Path[] segments, boolean normalize,
       LOG.info("CrawlDb update: Merging segment data into db.");
     }
 
+    FileSystem fs = crawlDb.getFileSystem(getConf());
+    Path outPath = FileOutputFormat.getOutputPath(job);
     try {
-      int complete = job.waitForCompletion(true)?0:1;
+      boolean success = job.waitForCompletion(true);
+      if (!success) {
+        String message = "Crawl job did not succeed, job status:"
+            + job.getStatus().getState() + ", reason: "
+            + job.getStatus().getFailureInfo();
+        LOG.error(message);
+        cleanupAfterFailure(outPath, lock, fs);
+        throw new RuntimeException(message);
+      }
     } catch (IOException | InterruptedException | ClassNotFoundException e) {
-      FileSystem fs = crawlDb.getFileSystem(getConf());
-      LockUtil.removeLockFile(fs, lock);
-      Path outPath = FileOutputFormat.getOutputPath(job);
-      if (fs.exists(outPath))
-        fs.delete(outPath, true);
+      LOG.error("Crawl job failed ", e);
```

Comment:
Please always use parameterized logging e.g.
```LOG.error("Crawl job failed {}", e);```

Reply:
ACK! I will incorporate this change and send a commit covering all the 25 instances. Thanks.

Reply:
+1

The CrawlDb cleanup method looks same as that from Injector: maybe move it to CrawlDb - there are already methods to lock and install the CrawlDb shared by various jobs (inject, update, merge, generate).

Reply:
Why not shift the sub routine cleanup to the class nutch/src/java/org/apache/nutch/util/NutchJob.java? This is a nutch job utils file, I think this would be more appropriate. I have cross checked that cleanup sub routine would be useful in outside crawld like in hostdb, indexer etc.